functional: more fleetctl unit action tests

dongsupark opened the pull request:
In order to make functional tests cover more fleetctl commands, do the following:

* define ``util.UnitFileState`` for handling output of list-unit-files
* aadd new cluster operations ``WaitForNUnits`` and ``WaitForNUnitFiles``, wrappers based on ``util.WaitForState``.
* introduce a new test ``TestUnitCat`` for "``fleetctl cat``".
* introduce a new test ``TestUnitStatus`` for "``fleetctl status``".
* improve an existing test ``TestUnitSubmit``.
* introduce a new test ``TestUnitLoad`` for ``fleetctl load``.

tixxdz commented on functional/unit_action_test.go:
```diff
@@ -73,48 +76,47 @@ func TestUnitSubmit(t *testing.T) {
+	if err != nil {
 		t.Fatalf("Unable to submit fleet unit: %v", err)
 	}
-	stdout, _, err := cluster.Fleetctl(m, "list-units", "--no-legend")
```
@dongsupark thanks for the patches!

Hmm something is wrong here, my understanding with fleet is submit only sends the units into the cluster, so it will show up when you do "fleetctl list-unit-files" but not with "fleetctl list-units" since the unit is not active.

Was this test wrong from the beginning ? we should not check the number of printed lines, we should just grep for the name of the unit, thoughts ?

Thanks!

dongsupark replied:
@tixxdz Yes, this part has been wrong since the day 0. (5f1340b3)
I'm not sure how the test has always succeeded. I'll try to improve it.

jonboulle replied:
It wasn't "wrong", just not specific enough, right?

tixxdz replied:
@jonboulle yes! by "wrong" I'm only referring to the functional tests and that perhaps there is a bug somewhere that allowed it to pass. For fleetctl commands they are clear.

tixxdz replied:
@jonboulle yes not specific enough ;-)

tixxdz replied:
@dongsu capturing the context, maybe we need a function `runCheckListUnits()` , `runCheckListUnitFiles()` and if you want to grep submitted units since they won't show up in other commands and you are not sure about the number of returned units... then make runGetListUnitFiles() return the units and grep it into the caller ? just quick thoughts.

thanks

antrik replied:
@jonboulle no, it was indeed completely wrong: `list-units` returns an empty string; but that's still counts as one line, so the test "passed"...

antrik replied:
Note that most of the existing tests only check the number of units, except for cases where there is actually a possibility for ambiguity -- so checking only the count would be fine here as well. It just needs to be fixed to use `list-unit-files`; and to properly check for a non-empty string...

antrik replied:
I just saw that one of the existing tests in this very file (namely `TestUnitRunnable()`) actually test both the count *and* the name -- so feel free to copy that if you really think it's worthwhile...

dongsupark replied:
@antrik 
> no, it was indeed completely wrong: list-units returns an empty string; but that's still counts as one line, so the test "passed"...

Yes, that was exactly what I wanted to say. That's why I said yesterday that this part has been wrong since the beginning. I think the underlying reason was that ``len(strings.Split("", "\n"))`` returns 1, not 0 as expected.

So my solution to get it working was adding additional conditions like "``expectedUnits == 0 && len(outListUnitFiles) == 0``" in each check function. Still somehow nasty, but it's working for now. Afterwards, we should do some code auditing in the entire tree to find such errors. But anyway that would be over the scope of this PR.

dongsupark commented:
Updated.
* add new utilities ``util.SliceContains`` and ``util.GetListUnitNames``.
* add a new helper ``runListUnitFiles``, a wrapper for ``fleetctl list-unit-files``.
* improve ``TestUnitSubmit`` to correctly check the output running by list-unit-files instead of list-units, as well as to parse output to compare the unit name with the expected name.
* improve ``TestUnitLoad`` to parse output to compare the unit name with the expected name.

antrik commented on functional/unit_action_test.go:
```diff
@@ -69,47 +76,132 @@ func TestUnitSubmit(t *testing.T) {
 	// submit a unit and assert it shows up
-	if _, _, err := cluster.Fleetctl(m, "submit", "fixtures/units/hello.service"); err != nil {
+	_, _, err = cluster.Fleetctl(m, "submit", unitFile)
+	if err != nil {
```
Why did you change the syntax here? There is no reason to have the assignment outside the `if`.

antrik commented on functional/unit_action_test.go:
```diff
@@ -69,47 +76,132 @@ func TestUnitSubmit(t *testing.T) {
+		t.Fatalf("Unable to load fleet unit: %v", err)
+	}
+
+	// wait until the unit gets unloaded up to 15 seconds
```
s/unloaded/loaded/

antrik replied:
This one is still wrong too.

antrik commented on functional/unit_action_test.go:
```diff
@@ -224,3 +316,139 @@ func TestUnitSSHActions(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	// submit a unit and assert it shows up
```
Eh?

antrik commented on functional/unit_action_test.go:
```diff
@@ -224,3 +316,139 @@ func TestUnitSSHActions(t *testing.T) {
+// runListUnits runs fleetctl list-units to verify the actual number of units
+// matched with the given expected number. It periodically runs list-units
+// waiting until list-units actually shows the expected units.
+func runListUnits(cluster platform.Cluster, m platform.Member, expectedUnits int) ([]string, error) {
```
Can't you just use the `WaitForNUnits()` function you introduced in another PR? Seems to do the same more or less...

tixxdz replied:
That's not a good idea we explicitly want to know from a caller perspective that this command is running `fleetctl list-units` so we use it where it's appropriate.

antrik commented on functional/unit_action_test.go:
```diff
@@ -224,3 +316,139 @@ func TestUnitSSHActions(t *testing.T) {
+// of units matched with the given expected number. It periodically runs
+// list-unit-files waiting until list-unit-files actually shows the expected
+// number of unit files.
+func runListUnitFiles(cluster platform.Cluster, m platform.Member, expectedUnits int) ([]string, error) {
```
For consistency, this probably should be `WaitForNUnitFiles()`, with similar semantics to `WaitForNUnits()` and `WaitForNActiveUnits()`.

tixxdz replied:
Same thing here, I've a functional test waiting #1515 that will use the runListUnitFiles() which is `fleetctl list-unit-files` and I want the returned output, so I could do the parsing. So WaitForNUnit() should be either inside these functions or just on its own. Functional tests should just be easy to write I want to run this function and get its output...

antrik commented on functional/util/util.go:
```diff
@@ -154,3 +154,32 @@ func WaitForState(stateCheckFunc func() bool) (time.Duration, error) {
+
+// SliceContains returns true if every string in needle slice exists in any
+// slice in haystack.
+func SliceContains(haystack []string, needle []string) bool {
```
You should probably check that the list of units is *identical* to the expected ones, not that it's a superset...

But for the actual test cases in question, it's probably overkill anyway -- see my first comment, about just checking the count...

antrik commented on functional/util/util.go:
```diff
@@ -154,3 +154,32 @@ func WaitForState(stateCheckFunc func() bool) (time.Duration, error) {
+
+// GetListUnitNames returns the 1st field of the given string, which is
+// a single line of output of "fleetctl list-{units,unit-files}".
+func GetListUnitNames(listUnitFiles []string) []string {
```
Note that when you go with a `WaitForNUnits()` and `WaitForNUnitFiles()` along the lines of the existing `WaitForNActiveUnits()` -- i.e. returning a map -- you won't need this helper...

tixxdz replied:
That's a good point! so since it's just functional maybe return a map unless some other points that we are not aware of...

tixxdz commented:
So a couple of changes then maybe merge this one ?

dongsupark commented:
Updated.
* define ``util.UnitFileState`` for handling output of list-unit-files
* add new cluster operations ``WaitForNUnits`` and ``WaitForNUnitFiles``
* make ``TestUnitSubmit()`` use ``cluster.WaitForNUnitFiles``
* make ``TestUnitLoad()`` use ``cluster.WaitForNUnits``
* fix a bug in ``TestUnitStatus()``, loading a unit before getting status.
* drop unnecessary helpers like ``SliceContains``, ``runListUnits``, etc.
* fix minor things like comments or coding style

I assume this PR should be merged prior to the existing #1509.
So I might also have to update functional tests in #1509 to avoid conflicts.

antrik commented on functional/unit_action_test.go:
```diff
@@ -69,47 +74,133 @@ func TestUnitSubmit(t *testing.T) {
 		t.Fatalf("Failed to destroy unit: %v", err)
 	}
-	stdout, _, err = cluster.Fleetctl(m, "list-units", "--no-legend")
+	// wait until the unit gets submitted up to 15 seconds
```
I think this comment is still wrong here...

antrik commented on functional/unit_action_test.go:
```diff
@@ -224,3 +315,97 @@ func TestUnitSSHActions(t *testing.T) {
+		t.Fatalf("Unable to load a fleet unit: %v", err)
+	}
+
+	// wait until the unit gets unloaded up to 15 seconds
```
And another wrong comment...

antrik commented on functional/unit_action_test.go:
```diff
@@ -224,3 +315,97 @@ func TestUnitSSHActions(t *testing.T) {
+		t.Fatalf("Failed to run list-units: %v", err)
+	}
+
+	// NOTE: cluster.Fleetctl() cannot be used due to endpoint being messed up.
```
Huh? I don't understand this comment...

dongsupark replied:
@antrik Aha, this was my mistake. I fixed it into using ``cluster.Fleetctl()``. Thanks.

antrik commented on functional/platform/nspawn.go:
```diff
@@ -128,6 +128,46 @@ func (nc *nspawnCluster) FleetctlWithInput(m Member, input string, args ...strin
+		}
+		if expectedUnits == 0 && len(outListUnits) == 0 {
+			return true
+		}
```
Note that `ParseUnitStates()` implicitly filters out an empty line -- so the explicit check shouldn't be necessary here.

dongsupark replied:
@antrik 
> Note that ParseUnitStates() implicitly filters out an empty line

Yeah, I just got cold feet. Actually I would not like to rely on such implicit behavior. Anyway I removed checking for empty line, and wrote a comment about that.

antrik commented on functional/platform/nspawn.go:
```diff
@@ -165,6 +205,48 @@ func (nc *nspawnCluster) WaitForNActiveUnits(m Member, count int) (map[string][]
+		}
+		if expectedUnits == 0 && len(outListUnitFiles) == 0 {
+			return true
+		}
```
Same as above: no need to special-case `0` here.

dongsupark commented:
Done.
* Fixed a bug in ``TestUnitStatus``, by calling ``cluster.Fleetctl()`` correctly.
* Removed unnecessary checking in ``cluster.WaitForN{Units,UnitFiles}()``
* Fixed comments

tixxdz commented:
lgtm! thanks all.
